### PR TITLE
GUI2/Game Load: refactor filter implementation

### PR DESCRIPTION
**Pull request by Vultraz**
Opening this for feedback. This replaces the old filter code with a new regex-based method, and adds a new function to filter listboxes based on a predicate instead of providing a mask directly.

In terms of performance, it's extremely fast on my machine (often under 1ms). It pulls in the `fmt` library as a substitute for `<format>` (there's a flag I can set for header-only mode), which could be further deployed elsewhere in preparation for the standard library version once we can use C++20.

### CODE REVIEW

**Comment by soliton-** [comment]
> In terms of performance, it's extremely fast on my machine (often under 1ms).

Compared to what?

**Comment by stevecotton** [comment]
> It pulls in the `fmt` library as a substitute for `<format>` (there's a flag I can set for header-only mode), which could be further deployed elsewhere in preparation for the standard library version once we can use C++20.

Why would we want to use `std::format`? I can see its value to projects that have lots of old `printf` statements, but in Wesnoth there are only a handful in `src/attack_prediction.cpp` and `src/conftests/sdl2_audio.cpp`. For those two files it might be nice cleanup once we move to C++20, but it doesn't seem a reason to rush to use `fmt` now.

For reordering the arguments in translatable strings, `std::format` only supports numerical placeholders, so `_("$game_name (Era: $era_name)")` would have to become `_("{1} (Era: {2})")`. I don't see a case where we'd replace `VGETTEXT` with that.

File: src/gui/dialogs/game_load.cpp [hunk 1]
```diff
@@ -314,40 +316,15 @@ void game_load::display_savegame()
 
 void game_load::filter_text_changed(const std::string& text)
 {
-	listbox& list = find_widget<listbox>(get_window(), "savegame_list", false);
-
-	const std::vector<std::string> words = utils::split(text, ' ');
+	auto split = utils::split(text, ' ');
+	std::transform(split.begin(), split.end(), split.begin(),
+		[](std::string_view str) { return fmt::format("(?=.*{})", str); });
```
[the remark this answers is not in the document]
Comment (CelticMinstrel) [reply src/gui/dialogs/game_load.cpp:321]: Yeah, this is no good. It's a potential crash, isn't it? Just try filtering for, say, `{` or `?`.

While there might be an argument for allowing people to use regex in the filter box, I think it's probably better to escape the special characters.

**Comment by CelticMinstrel** [comment]
> Why would we want to use `std::format`?

Glancing through the source, I can see some uses of `formatter()` that should probably be replaced with a format string, so that could be an answer… although for the most part we have our custom formatting function that can be used instead. Still, that function doesn't support formatting parameters like precision and the like. If `std::format` truly supports only numerical prefixes, and we need to be able to control precision, it's probably easier to actually _add_ support for that sort of thing to our custom format syntax.

**Comment by stevecotton** [comment]
> If `std::format` truly supports only numerical prefixes, and we need to be able to control precision, it's probably easier to actually _add_ support for that sort of thing to our custom format syntax.

It does provide [formatting for precision](https://en.cppreference.com/w/cpp/utility/format/formatter#Standard_format_specification) too, I was only referring to the requirement of reordering arguments in translatable strings.

So okay, replacing `formatter()` with `std::format` does make sense once we've moved to C++20. I still don't see a reason to move to the `fmt` library before we can move to `std::format` itself.

**Comment by Pentarctagon** [comment]
> If we wanted log messages to be translatable, std::format would be essential, but I believe the consensus is that logs shouldn't be translated? Though I'm not sure if that even makes sense.

I'd really rather not have log messages be subject to string freezes.

**Comment by Vultraz** [comment]
> Why would we want to use `std::format`?
There are many, many cases in the codebase of manual string formatting vs concatenation, especially for constructing pango markup strings in the UI. Besides that, as mentioned, we have the `formatter()` class. I'd eventually like to clean all those up.

File: src/gui/dialogs/game_load.cpp (hunk 1 above)
[the remark this answers is not in the document]
Comment (Vultraz) [reply src/gui/dialogs/game_load.cpp:321]: Yeah, it's a crash. Pondering how best to handle.

**Comment by gfgtdf** [comment]
If I read the reflex correctly the main difference in behavior is that a search now only matches if the terms appear in the correct order? I don't really have an opinion on whether this is good or not, might actually be an advantage, not sure.


On a technical level we either need to escape the input string before creating a reflex with it or/and catch bad_regex exception to prevent errors. I do think that at least the ability to match different word "or"-wise could actually be useful for users though, so maybe don't escape `|` then ?

**Comment by CelticMinstrel** [comment]
> > If `std::format` truly supports only numerical prefixes, and we need to be able to control precision, it's probably easier to actually _add_ support for that sort of thing to our custom format syntax.
> 
> It does provide [formatting for precision](https://en.cppreference.com/w/cpp/utility/format/formatter#Standard_format_specification) too, I was only referring to the requirement of reordering arguments in translatable strings.

I was referring to that too. I think I wrote "prefixes" where I meant to write "placeholders". But what I'm saying is that if we _need_ precision specifiers, and we _also need_ reorderable arguments, it would make sense to add precision specifiers to the $-syntax. I think copying the way Unreal Engine handles argument modifiers would fit well in that syntax, but use ~ to evoke the IPF syntax. So you'd write `$stuff|~precision(.2)` or something (not sure exactly, but along those lines).

> I'd really rather not have log messages be subject to string freezes.

If we wanted them translatable for some reason, we could always put them in a separate low-priority non-frozen textdomain. However, the question is, who is the target audience for log messages? I'd argue the answer to that is us, in which case translating them is pointless. (Which also means we shouldn't be putting stuff in log messages that we expect actual users would want to see.)

> There are many, many cases in the codebase of manual string formatting via concatenation, especially for constructing pango markup strings in the UI. Besides that, as mentioned, we have the `formatter()` class. I'd eventually like to clean all those up.

Constructing through concatenation or `formatter()` is fine if it's not for UI, but there are definitely some uses of `formatter()` in UI that we may wish to clean up. They mainly seem to be constructing paragraphs or "key: value" pairs with intervening Pango markup, so it's probably not a major issue for translations, but it would still be easier to read if it used format strings to some extent.

> If I read the reflex correctly the main difference in behavior is that a search now only matches if the terms appear in the correct order? I don't really have an opinion on whether this is good or not, might actually be an advantage, not sure.

I think you're not reading correctly. Vultraz appears to be aiming for an "intersection type" by using a series of "lookahead assertions". So it'll match the full text against each word in turn (with the leading `.*`) until it finds one that doesn't match, at which point it will fail. If it doesn't find one that doesn't match, then it passes. So the words can appear in any order.

> On a technical level we either need to escape the input string before creating a reflex with it or/and catch bad_regex exception to prevent errors. I do think that at least the ability to match different word "or"-wise could actually be useful for users though, so maybe don't escape `|` then ?

Basically this. Blindly dumping user input into a regex is _not_ the right answer. I don't have a strong opinion on whether or not to escape `|`, or even on whether to escape or catch exceptions. But you need to do _something_ more than what you're doing.

**Comment by stevecotton** [comment]
Something in the UI ought to tell the user what patterns they can use, which is hard when having to explain regexs.
